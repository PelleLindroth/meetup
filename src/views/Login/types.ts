import { User } from '../../db/models/User'

export type LoginProps = {
  setUser: React.Dispatch<React.SetStateAction<User | null>>
}
