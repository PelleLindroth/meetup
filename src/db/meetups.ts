import { User, users } from './users'

export interface Meetup {
  id: string
  title: string
  description?: string
  date: Date
  location?: Address
  url?: string
  online: boolean
  capacity: number | null
  attending: number
  arranger: User
  reviews: Review[]
  comments: Comment[]
}

interface Address {
  street: string
  city: string
  country: string
}

interface Review {
  id: string
  meetupId: string
  rating: number
  submittedAt: Date
}

interface Comment {
  id: string
  meetupId: string
  title: string
  body: string
  submittedAt: Date
}

export const meetups: Meetup[] = [
  {
    id: '1',
    title: 'Pizza Picnic',
    description: 'We meet in the park with pizzas and hang out all night',
    date: new Date(1630170000000),
    location: {
      street: 'Vitabergsparken',
      city: 'Stockholm',
      country: 'Sweden',
    },
    online: false,
    url: 'https://www.udemy.com/course/rust-lang/',
    capacity: null,
    attending: 54,
    arranger: users[0],
    reviews: [],
    comments: [],
  },
  {
    id: '2',
    title: 'Rust course',
    description: 'Learn a new programming language from the ground up',
    date: new Date(1641808800000),
    online: true,
    url: 'https://www.udemy.com/course/rust-lang/',
    capacity: null,
    attending: 234,
    arranger: users[2],
    reviews: [],
    comments: [],
  },
  {
    id: '3',
    title: 'Sevilla - Real Sociedad',
    description: 'A classic football match between the south and the north',
    date: new Date(1652457600000),
    location: {
      street: 'C. Sevilla Fútbol Club',
      city: '41005 Sevilla',
      country: 'Spain',
    },
    online: false,
    capacity: 43883,
    attending: 23478,
    arranger: users[1],
    reviews: [],
    comments: [],
  },
  {
    id: '4',
    title: 'Kite festival',
    description:
      'Classic Kite festival at Gärdet, bring your own kite or just watch the pros',
    date: new Date(1656774000000),
    location: {
      street: 'Gärdet',
      city: 'Stockholm',
      country: 'Sweden',
    },
    online: false,
    capacity: null,
    attending: 126,
    arranger: users[3],
    reviews: [],
    comments: [],
  },
]
