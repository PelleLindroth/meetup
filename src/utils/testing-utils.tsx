import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Routes, Route } from 'react-router'

export const renderWithRouter = (ui: ReactElement, options?: any) =>
  render(ui, { wrapper: MemoryRouter, ...options })

export const renderWithPath = (ui: ReactElement, entry: string, path: string) =>
  render(
    <MemoryRouter initialEntries={[entry]}>
      <Routes>
        <Route path={path} element={ui} />
      </Routes>
    </MemoryRouter>
  )
