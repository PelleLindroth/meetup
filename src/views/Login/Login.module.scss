.login {
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-top: 5rem;
  text-align: center;

  h1 {
    color: #333;
    text-transform: uppercase;
  }

  form {
    align-items: center;
    background-color: #dcddde;
    border-radius: 0.4rem;
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin: auto;
    padding: 4rem 1rem;
    max-width: 20rem;
    width: 75%;

    input {
      background-color: #fff;
      border: none;
      border-radius: 2px;
      font-size: 1rem;
      outline: none;
      margin: 0.5rem 0;
      padding: 0.5rem;
      width: 85%;
    }

    .errorWrapper {
      height: 2rem;

      .errorMessage {
        color: #c80000;
      }
    }
    button {
      background-color: #14c747;
      border: 1px solid transparent;
      border-radius: 4px;
      color: #fff;
      cursor: pointer;
      margin-top: 3rem;
      outline: none;
      padding: 0.4rem 1.7rem;
      text-transform: uppercase;
    }

    button:disabled {
      background-color: transparent;
      border: 1px solid #444;
      color: #444;
    }
  }
}
