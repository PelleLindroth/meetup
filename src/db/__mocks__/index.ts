import { meetups, Meetup } from '../meetups'
import { users, User } from '../users'
import { sortMeetupsChronologically } from '../dbutils'

export const getAllMeetups = () => {
  sortMeetupsChronologically(meetups)

  return meetups
}

export const getMeetupById = (id: string): Meetup | undefined => {
  return meetups.find((meetup) => meetup.id === id)
}

export const addComment = jest.fn()

export const addReview = jest.fn()

export const addMeetup = (meetup: Meetup) => {
  meetups.push(meetup)
}

export const validateUser = (email: string, password: string): User | null => {
  const user = users.find((user) => user.email === email)

  if (user) {
    return user.password === password ? user : null
  } else {
    return null
  }
}

export const signUpForEvent = (meetup: Meetup, user: User) => {
  user.attending.push(meetup.id)
  meetup.attending++
}

export const cancelSignUpForEvent = (meetup: Meetup, user: User) => {
  user.attending = user.attending.filter((item) => item !== meetup.id)
  meetup.attending--
}

export const storeUser = jest.fn()

export const getStoredUser = jest.fn()

export const clearStoredUser = jest.fn()

export const getUserById = (id: string): User | undefined => {
  return users.find((user) => user.id === id)
}

export const getUsers = () => {
  return users
}

export const storeUserAttending = jest.fn()

export const getUserAttending = (userId: string): string[] => {
  return getUserById(userId)!.attending
}

export const storeUserReviewed = jest.fn()

export const getUserReviewed = (userId: string): string[] => {
  return getUserById(userId)!.reviewed
}
