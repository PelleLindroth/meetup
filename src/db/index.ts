import { meetups, Meetup } from './meetups'
import { users, User } from './users'

export const getAllMeetups = (): Meetup[] => {
  const storedMeetups: Meetup[] = JSON.parse(localStorage.getItem('meetups')!)
  let returnedMeetups: Meetup[] = []

  if (storedMeetups) {
    returnedMeetups.push(...storedMeetups)
  } else returnedMeetups.push(...meetups)

  returnedMeetups.sort((a, b) => {
    if (a.date.getTime() > b.date.getTime()) {
      return 1
    } else if (b.date.getTime() > a.date.getTime()) {
      return -1
    } else {
      return 0
    }
  })

  return returnedMeetups
}

export const getMockMeetups = () => {
  meetups.sort((a, b) => {
    if (a.date.getTime() > b.date.getTime()) {
      return 1
    } else if (b.date.getTime() > a.date.getTime()) {
      return -1
    } else {
      return 0
    }
  })

  return meetups
}

export const saveMeetupToLocalStorage = (meetup: Meetup) => {
  const storedMeetups: Meetup[] = JSON.parse(localStorage.getItem('meetups')!)

  if (storedMeetups) {
    storedMeetups.push(meetup)

    localStorage.setItem('meetups', JSON.stringify(storedMeetups))
  } else {
    localStorage.setItem('meetups', JSON.stringify([meetup]))
  }
}

export const getMeetupById = (id: string): Meetup | undefined => {
  return meetups.find((meetup) => meetup.id === id)
}

export const validateUser = (email: string, password: string): User | null => {
  const user = users.find((user) => user.email === email)

  if (user) {
    return user.password === password ? user : null
  } else {
    return null
  }
}

export const getUserById = (id: string): User | undefined => {
  return users.find((user) => user.id === id)
}

export const getMockUsers = () => {
  return users
}
