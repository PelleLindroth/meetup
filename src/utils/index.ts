import { User } from '../db/users'
import { Meetup } from '../db/meetups'

export const formatDate = (date: Date) => {
  return `${new Intl.DateTimeFormat('en-GB', {
    year: 'numeric',
    month: 'long',
    day: '2-digit',
    hour: 'numeric',
    minute: 'numeric',
  }).format(date)}`
}

const sortChronologically = (array: Meetup[]) => {
  return array.sort((a, b) => a.date.getTime() - b.date.getTime())
}

export const createMeetupList = (
  meetups: Meetup[],
  listType: string,
  searchPhrase: string,
  searchFilter: string
) => {
  const meetupList = meetups.filter((meetup) => {
    if (listType === 'past') {
      return (
        meetup.date.getTime() < Date.now() &&
        meetup.title.toLowerCase().includes(searchPhrase.toLowerCase()) &&
        (searchFilter === listType || searchFilter === 'all')
      )
    } else {
      return (
        meetup.date.getTime() > Date.now() &&
        meetup.title.toLowerCase().includes(searchPhrase.toLowerCase()) &&
        (searchFilter === listType || searchFilter === 'all')
      )
    }
  })

  return sortChronologically(meetupList)
}

export const createUserLists = (user: User, meetups: Meetup[]) => {
  return {
    ownUpcoming: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          meetup.arranger.id === user!.id &&
          meetup.date.getTime() > new Date().getTime()
      )
    ),
    ownPast: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          meetup.arranger.id === user!.id &&
          meetup.date.getTime() < new Date().getTime()
      )
    ),
    upcomingAttending: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          user!.attending.includes(meetup.id) &&
          meetup.date.getTime() > new Date().getTime()
      )
    ),
    pastAttended: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          user!.attending.includes(meetup.id) &&
          meetup.date.getTime() < new Date().getTime()
      )
    ),
  }
}
