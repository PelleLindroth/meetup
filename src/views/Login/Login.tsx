import { useState } from 'react'
import { useNavigate } from 'react-router'
import styles from './Login.module.scss'
import { User } from '../../db/users'
import { validateUser, storeUser, getUserDetails } from '../../db'

type LoginProps = {
  setUser: React.Dispatch<React.SetStateAction<User | null>>
}

const Login = (props: LoginProps) => {
  const { setUser } = props
  const [email, setEmail] = useState<string>('')
  const [password, setPassword] = useState<string>('')
  const [error, setError] = useState<boolean>(false)
  const navigate = useNavigate()

  const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault()
    setError(false)

    const user = validateUser(email, password)

    if (user) {
      const { attending, reviewed } = getUserDetails(user.id)!

      user.attending = [...attending]
      user.reviewed = [...reviewed]

      setUser(user)
      storeUser(user.id)
      navigate('/', { replace: true })
    } else {
      setError(true)
    }
  }

  return (
    <section className={styles.login}>
      <h1>Log in</h1>
      <form onSubmit={handleSubmit} name="login">
        <label htmlFor="email">Email</label>
        <input
          aria-label="email"
          autoComplete="current-email"
          id="email"
          value={email}
          onChange={(e) => setEmail(e.currentTarget.value)}
          type="email"
          placeholder="Email"
        />
        <label htmlFor="password">Password</label>
        <input
          aria-label="password"
          autoComplete="current-password"
          id="password"
          value={password}
          onChange={(e) => setPassword(e.currentTarget.value)}
          type="password"
          placeholder="Password"
        />
        <div className={styles.errorWrapper}>
          {error && (
            <p className={styles.errorMessage}>Invalid email or user name</p>
          )}
        </div>
        <button disabled={!email.length || !password.length}>Submit</button>
      </form>
    </section>
  )
}

export default Login
