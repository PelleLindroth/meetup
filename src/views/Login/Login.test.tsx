import { screen } from '@testing-library/react'
import { mountWithRouter, renderWithRouter } from '../../utils/testing-utils'
import Login from './index'
import userEvent from '@testing-library/user-event'
import { getMockUsers } from '../../db'

describe('Login Unit Tests', () => {
  it('renders login form with two empty inputs and a button', () => {
    const wrapper = mountWithRouter(<Login setUser={jest.fn()} />)

    expect(wrapper.find(Login)).toMatchSnapshot()
    expect(wrapper.find(Login).find('input').at(0).text()).toBe('')
    expect(wrapper.find(Login).find('input').at(1).text()).toBe('')
    expect(wrapper.find(Login).find('button').text()).toBe('Submit')
  })
  it('disables login button by default', () => {
    renderWithRouter(<Login setUser={jest.fn()} />)

    const loginButton = screen.getByRole('button', { name: /submit/i })

    expect(loginButton).toBeDisabled()
  })
  it('enables login button when both input fields have text', () => {
    renderWithRouter(<Login setUser={jest.fn()} />)

    const emailInput = screen.getByRole('textbox', { name: /email/i })
    const passwordInput = screen.getByLabelText(/password/i)
    const loginButton = screen.getByRole('button', { name: /submit/i })

    userEvent.type(emailInput, 'tjalle')
    userEvent.type(passwordInput, 'grillkorv')

    expect(loginButton).toBeEnabled()
  })
  it('calls setUser once with user object if credentials are valid', () => {
    const setUserSpy = jest.fn()
    const mockUsers = getMockUsers()

    renderWithRouter(<Login setUser={setUserSpy} />)

    const emailInput = screen.getByRole('textbox', { name: /email/i })
    const passwordInput = screen.getByLabelText(/password/i)
    const loginButton = screen.getByRole('button', { name: /submit/i })

    userEvent.type(emailInput, mockUsers[0].email)
    userEvent.type(passwordInput, mockUsers[0].password)
    userEvent.click(loginButton)

    expect(setUserSpy).toHaveBeenCalledTimes(1)
    expect(setUserSpy).toHaveBeenCalledWith(mockUsers[0])
  })
  it('displays an error message if credentials are invalid', () => {
    renderWithRouter(<Login setUser={jest.fn()} />)

    const emailInput = screen.getByRole('textbox', { name: /email/i })
    const passwordInput = screen.getByLabelText(/password/i)
    const loginButton = screen.getByRole('button', { name: /submit/i })

    userEvent.type(emailInput, 'rolle@ica.se')
    userEvent.type(passwordInput, 'palsternacka123')
    userEvent.click(loginButton)

    const errorMessage = screen.getByText(/invalid/i)

    expect(errorMessage).toBeInTheDocument()
  })
})
