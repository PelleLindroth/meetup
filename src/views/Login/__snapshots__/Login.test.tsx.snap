// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Unit Tests renders login form with two empty inputs and a button 1`] = `
<Login
  setUser={[MockFunction]}
>
  <section
    className="login"
  >
    <h1>
      Log in
    </h1>
    <form
      name="login"
      onSubmit={[Function]}
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <input
        autoComplete="current-password"
        id="email"
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        autoComplete="current-password"
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <div
        className="errorWrapper"
      />
      <button
        disabled={true}
      >
        Submit
      </button>
    </form>
  </section>
</Login>
`;
